src/Dhall/Parser/Vector.hs

```diff
@@ -0,0 +1,199 @@
+{-# LANGUAGE MultiWayIf          #-}
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies        #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+-- | Unboxed vector as megaparsec's Stream.
+module Dhall.Parser.Vector (
+    UVectorChar,
+    uvectorToText,
+    uvectorFromText,
+    ) where
+
+import Data.Proxy (Proxy (..))
+import Text.Megaparsec (SourcePos (..), PosState (..), Token, Tokens, pos1, mkPos, unPos, takeWhile_)
+
+import qualified Text.Megaparsec
+import qualified Data.Text
+import qualified Data.Vector.Unboxed as U
+
+type UVectorChar =  U.Vector Char
+
+uvectorToText :: UVectorChar -> Data.Text.Text
+uvectorToText = Data.Text.pack . U.toList
+
+uvectorFromText :: Data.Text.Text -> UVectorChar
+uvectorFromText = U.fromList . Data.Text.unpack
+
+instance c ~ Char => Text.Megaparsec.Stream (U.Vector c) where
+    type Token (U.Vector c) = c
+    type Tokens (U.Vector c) = U.Vector c
+
+    tokenToChunk _  = U.singleton
+    tokensToChunk _ = U.fromList
+    chunkToTokens _ = U.toList
+    chunkLength _   = U.length
+    chunkEmpty _    = U.null
+
+    take1_ xs
+        | U.null xs = Nothing
+        | otherwise = Just (U.unsafeHead xs, U.unsafeTail xs)
+
+    takeN_ n xs
+        | n <= 0           = Just (U.empty, xs)
```

Comment:
Is it faster to omit the special-case logic for `n <= 0`?  The reason I ask is that in the common case `n` will be greater than `0` so this is adding another predicate that will have to be checked every time we parse a chunk of fixed length

Same question about the `U.length xs <= n` predicate, which will return `False` the vast majority of the time since most requested chunks will just be a few characters (i.e. `Natural/fold` or `++`) and the input code will usually be long (i.e. kilobytes minimally) and whenever the input code is short the parsing performance doesn't matter :)

Reply:
good points. These are somewhat direct transliteration of `Text` instance, but it make sense to experiment with clause ordering